src/main/java/com/exasol/spark/s3/S3Source.java

```diff
@@ -0,0 +1,154 @@
+                            .mitigation("Please use only either one of the options.").toString());
+        }
+        validateRequiredOptions(options);
+    }
+
+    private void validateRequiredOptions(CaseInsensitiveStringMap options) {
+        for (final String key : REQUIRED_OPTIONS) {
+            if (!options.containsKey(key)) {
+                throw new IllegalArgumentException(ExaError.messageBuilder("E-SEC-14")
+                        .message("Required option {{KEY}} is not found.")
+                        .mitigation("Please provide a value for the {{KEY}} option.").parameter("KEY", key).toString());
+            }
+        }
+    }
+
+    private ExasolOptions getExasolOptions(final CaseInsensitiveStringMap options) {
+        final ExasolOptions.Builder builder = ExasolOptions.builder() //
+                .jdbcUrl(options.get(JDBC_URL)) //
+                .username(options.get(USERNAME)) //
+                .password(options.get(PASSWORD));
+        if (options.containsKey(TABLE)) {
+            builder.table(options.get(TABLE));
+        } else if (options.containsKey(QUERY)) {
+            builder.query(options.get(QUERY));
+        }
+        return builder.build();
+    }
+
+    private StructType getSchema(final ExasolOptions options) {
+        final String limitQuery = generateInferSchemaQuery(options);
+        LOGGER.info(() -> "Running schema inference using limited query '" + limitQuery + "' for the s3 source.");
+        try (final Connection connection = new ExasolConnectionFactory(options).getConnection();
+                final Statement statement = connection.createStatement()) {
+            final StructType schema = getSparkSchema(statement.executeQuery(limitQuery));
+            LOGGER.info(() -> "Inferred schema as '" + schema.toString() + "' for the s3 source.");
+            return schema;
+        } catch (final SQLException exception) {
+            throw new ExasolConnectionException(ExaError.messageBuilder("E-SEC-15")
+                    .message("Could not run the limit query {{limitQuery}} to infer the schema.", limitQuery)
+                    .mitigation("Please check that connection properties and original query are correct.").toString(),
+                    exception);
+        }
+    }
+
+    private String generateInferSchemaQuery(final ExasolOptions options) {
+        final Select select = StatementFactory.getInstance().select();
+        select.all().from().table("<SCHEMA_INFERENCE_TABLE>");
+        if (options.hasQuery()) {
+            select.limit(1);
+        }
+        final StringRendererConfig rendererConfig = StringRendererConfig.builder().quoteIdentifiers(true).build();
+        final SelectRenderer renderer = new SelectRenderer(rendererConfig);
+        select.accept(renderer);
+        return renderer.render().replace("\"<SCHEMA_INFERENCE_TABLE>\"", getTableOrQuery(options));
```

Comment:
Yes, maybe it would be clearer. I opted with sql-statement-builder since didn't want to include SQL subquery as it is to protect against sql injections. Even though it will run on the users side.